File: src/main/java/org/capturecoop/cccolorutils/CCColorUtils.java
package org.capturecoop.cccolorutils;

import org.capturecoop.ccutils.utils.CCStringUtils;

import java.awt.*;
import java.awt.image.BufferedImage;

public class CCColorUtils {
    public enum DIRECTION {VERTICAL, HORIZONTAL}

    public static String rgb2hex(Color color) {
        return String.format("#%02x%02x%02x", color.getRed(), color.getGreen(), color.getBlue());
    }

    public static Color hex2rgb(String colorStr) {
        return new Color(Integer.valueOf(colorStr.substring(1, 3), 16), Integer.valueOf( colorStr.substring(3, 5), 16), Integer.valueOf(colorStr.substring(5, 7), 16));
    }

    public static BufferedImage createAlphaBar(Color color, int width, int height, DIRECTION direction) {
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = (Graphics2D) image.getGraphics();

        int amount = width;
        float step = 255F / width;
        float alpha = 0;

        if(direction == DIRECTION.VERTICAL) {
            amount = height;
            step = 255F / height;
        }

        for(int pos = 0; pos < amount; pos++) {
            g.setColor(new Color(color.getRed(), color.getGreen(), color.getBlue(), (int)alpha));
            switch(direction) {
                case VERTICAL: g.drawLine(0, pos, width, pos); break;
                case HORIZONTAL: g.drawLine(pos, 0, pos, height); break;
            }
            alpha += step;
        }

        g.dispose();
        return image;
    }

    public static BufferedImage createHSVHueBar(int width, int height, DIRECTION direction) {
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = (Graphics2D) image.getGraphics();

        float hue = 0F;
        int amount = width;
        float step = 1F / width;

        if(direction == DIRECTION.VERTICAL) {
            amount = height;
            step = 1F / height;
        }

        for(int pos = 0; pos < amount; pos++) {
            g.setColor(new CCHSB(hue, 1F, 1F).toRGB());
            switch(direction) {
                case VERTICAL: g.drawLine(0, pos, width, pos); break;
                case HORIZONTAL: g.drawLine(pos, 0, pos, height); break;
            }
            hue += step;
        }

        g.dispose();
        return image;
    }

    public static BufferedImage createHSVBox(int width, int height, float hue) {
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = (Graphics2D) image.getGraphics();

        float brightness = 1F;
        float saturation;
        float stepHeight = 1F / height;
        float stepWidth = 1F / width;

        for(int y = 0; y < height; y++) {
            saturation = 0;
            for(int x = 0; x < width; x++) {
                saturation += stepWidth;
                g.setColor(new CCHSB(hue, saturation, brightness).toRGB());
                g.drawLine(x, y, x, y);
            }
            brightness -= stepHeight;
        }

        g.dispose();
        return image;
    }

    public static String toStringColor(Color c) {
        if(c == null)
            return null;
        return CCStringUtils.format("Color[r: %c, g: %c, b: %c, a: %c]", c.getRed(), c.getGreen(), c.getBlue(), c.getAlpha());
    }

    public static Color setColorRed(Color c, int red) {
        return new Color(red, c.getGreen(), c.getBlue(), c.getAlpha());
    }

    public static Color setColorGreen(Color c, int green) {
        return new Color(c.getRed(), green, c.getBlue(), c.getAlpha());
    }

    public static Color setColorBlue(Color c, int blue) {
        return new Color(c.getRed(), c.getGreen(), blue, c.getAlpha());
    }

    public static Color setColorAlpha(Color c, int alpha) {
        return new Color(c.getRed(), c.getGreen(), c.getBlue(), alpha);
    }

    public static void drawRect(Graphics g, Rectangle rect) {
        g.drawRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static void fillRect(Graphics g, Rectangle rect) {
        g.fillRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static Color getContrastColor(Color color) {
        double y = (299f * color.getRed() + 587 * color.getGreen() + 114 * color.getBlue()) / 1000;
        return y >= 128 ? Color.black : Color.white;
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/CCColorUtils.kt
package org.capturecoop.cccolorutils

import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.Rectangle
import java.awt.image.BufferedImage

object CCColorUtils {
    fun rgb2hex(color: Color) = String.format("#%02x%02x%02x", color.red, color.green, color.blue)

    fun hex2rgb(colorStr: String) = try {
        val r = Integer.valueOf(colorStr.substring(1, 3), 16)
        val g = Integer.valueOf(colorStr.substring(3, 5), 16)
        val b = Integer.valueOf(colorStr.substring(5, 7), 16)
        Color(r, g, b)
    } catch (exception: Exception) {
        null
    }

    fun createAlphaBar(color: Color, width: Int, height: Int, direction: DIRECTION): BufferedImage {
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        val g = image.graphics as Graphics2D
        var amount = width
        var step = 255f / width
        var alpha = 0f
        if (direction == DIRECTION.VERTICAL) {
            amount = height
            step = 255f / height
        }
        for (pos in 0 until amount) {
            g.color = Color(color.red, color.green, color.blue, alpha.toInt())
            when (direction) {
                DIRECTION.VERTICAL -> g.drawLine(0, pos, width, pos)
                DIRECTION.HORIZONTAL -> g.drawLine(pos, 0, pos, height)
            }
            alpha += step
        }
        g.dispose()
        return image
    }

    fun createHSVHueBar(width: Int, height: Int, direction: DIRECTION): BufferedImage {
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val g = image.graphics as Graphics2D
        var hue = 0f
        var amount = width
        var step = 1f / width
        if (direction == DIRECTION.VERTICAL) {
            amount = height
            step = 1f / height
        }
        for (pos in 0 until amount) {
            g.color = CCHSB(hue, 1f, 1f).asColor()
            when (direction) {
                DIRECTION.VERTICAL -> g.drawLine(0, pos, width, pos)
                DIRECTION.HORIZONTAL -> g.drawLine(pos, 0, pos, height)
            }
            hue += step
        }
        g.dispose()
        return image
    }

    fun createHSVBox(width: Int, height: Int, hue: Float): BufferedImage {
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val g = image.graphics as Graphics2D
        var brightness = 1f
        var saturation: Float
        val stepHeight = 1f / height
        val stepWidth = 1f / width
        for (y in 0 until height) {
            saturation = 0f
            for (x in 0 until width) {
                saturation += stepWidth
                g.color = CCHSB(hue, saturation, brightness).asColor()
                g.drawLine(x, y, x, y)
            }
            brightness -= stepHeight
        }
        g.dispose()
        return image
    }

    fun toStringColor(c: Color?) = if (c == null) null else "Color[r: ${c.red}, g: ${c.green}, b: ${c.blue}, a: ${c.alpha}]"

    fun setColorRed(c: Color, red: Int) = Color(red, c.green, c.blue, c.alpha)
    fun setColorGreen(c: Color, green: Int) = Color(c.red, green, c.blue, c.alpha)
    fun setColorBlue(c: Color, blue: Int) = Color(c.red, c.green, blue, c.alpha)
    fun setColorAlpha(c: Color, alpha: Int) = Color(c.red, c.green, c.blue, alpha)

    fun drawRect(g: Graphics, rect: Rectangle) = g.drawRect(rect.x, rect.y, rect.width, rect.height)
    fun fillRect(g: Graphics, rect: Rectangle) = g.fillRect(rect.x, rect.y, rect.width, rect.height)

    fun getContrastColor(color: Color): Color {
        val y = (299f * color.red + 587 * color.green + 114 * color.blue / 1000).toDouble()
        return if (y >= 128) Color.black else Color.white
    }

    enum class DIRECTION { VERTICAL, HORIZONTAL }
}